File: src/main/java/com/extendedae_plus/common/impl/guiSync/PacketStreamable.java
package com.extendedae_plus.common.impl.guiSync;

import appeng.menu.guisync.PacketWritable;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/// @see com.extendedae_plus.mixin.core.ae2.MixinSyncedField.MixinFieldCustom
public interface PacketStreamable extends PacketWritable {
    Map<Class<? extends PacketStreamable>, StreamCodec<RegistryFriendlyByteBuf, ? extends PacketStreamable>> registry = new HashMap<>();

    static <TField extends PacketStreamable> void register(
            Class<TField> clazzField, StreamCodec<RegistryFriendlyByteBuf, TField> streamCodec) {
        if (registry.put(clazzField, streamCodec) != null)
            throw new IllegalStateException("Duplicate field found for " + clazzField.getSimpleName());
    }

    @SuppressWarnings("rawtypes")
    static Optional<StreamCodec> getStreamCodec(Class<?> clazzField) {
        var streamCodec = registry.get(clazzField);
        if (PacketStreamable.class.isAssignableFrom(clazzField) && streamCodec == null)
            throw new IllegalStateException("Unregistered streamable object: " + clazzField.getSimpleName());
        return Optional.ofNullable(streamCodec);
    }

    @SuppressWarnings("unchecked")
    @Override
    default void writeToPacket(RegistryFriendlyByteBuf data) {
        PacketStreamable.getStreamCodec(this.getClass())
                .ifPresent(streamCodec -> streamCodec.encode(data, this));
    }
}

File: src/main/kotlin/com/extendedae_plus/common/impl/guiSync/PacketStreamable.kt
package com.extendedae_plus.common.impl.guiSync

import appeng.menu.guisync.PacketWritable
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import kotlin.reflect.KClass

/**
 * @see com.extendedae_plus.mixin.core.ae2.MixinSyncedField.MixinFieldCustom
 */
interface PacketStreamable : PacketWritable {
    override fun writeToPacket(data: RegistryFriendlyByteBuf) =
        getStreamCodec(this::class)?.encode(data, this) ?: Unit

    companion object {
        fun <TField : PacketStreamable> register(
            clazzField: KClass<TField>,
            streamCodec: StreamCodec<RegistryFriendlyByteBuf, TField>
        ) = check(
            registry.put(
                clazzField,
                streamCodec
            ) == null
        ) { "Duplicate field found for " + clazzField.simpleName }

        @JvmStatic
        @Suppress("unchecked_cast")
        fun getStreamCodec(clazzField: KClass<*>): StreamCodec<RegistryFriendlyByteBuf, Any>? {
            val streamCodec = registry[clazzField]
            check(!(PacketStreamable::class.java.isAssignableFrom(clazzField.java) && streamCodec == null)) {
                "Unregistered streamable object: " + clazzField.simpleName
            }
            return streamCodec as? StreamCodec<RegistryFriendlyByteBuf, Any>
        }

        val registry = HashMap<KClass<out PacketStreamable>, StreamCodec<RegistryFriendlyByteBuf, out PacketStreamable>>()
    }
}
